### PR TITLE
Add a test harness for UDFs

### PR DESCRIPTION
**Pull request by jklukas**
This is the first stage for addressing #141

With this change, we parse dependencies between UDFs so that each
UDF sql file can contain just a single UDF definition.
Furthermore, we allow additional SQL statements in the file which
will be treated as tests; they should call ERROR to cause
a test to fail.

Opening as a draft before converting all the content, so that it's easier
to make course corrections based on feedback.

### CODE REVIEW

File: tests/parse_udf.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+def read_udf_dir(d):
+    """
+    Read contents of d into RawUdf instances.
+    """
+    for root, dirs, files in os.walk(d):
+        for filename in files:
+            filepath = f"{root}/{filename}"
+            with open(filepath) as f:
+                yield RawUdf.from_file(filepath)
```
Comment (relud) [inline tests/parse_udf.py]: looks like opening `filepath` moved into `RawUdf.from_file`
```diff
-            filepath = f"{root}/{filename}"
-            with open(filepath) as f:
-                yield RawUdf.from_file(filepath)
+            yield RawUdf.from_file(f"{root}/{filename}")
```

File: tests/test_udfs.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+from . import parse_udf
+
+@pytest.fixture(scope="session")
+def bq():
+    return bigquery.Client()
+
+def test_udfs(bq):
+    for udf in parse_udf.parse_udf_dir('udf'):
+        for i, test in enumerate(udf.tests):
```
Comment (relud) [inline tests/test_udfs.py]: If you turn these loops into parameters they can be parallelized
```diff
-def test_udfs(bq):
-    for udf in parse_udf.parse_udf_dir('udf'):
-        for i, test in enumerate(udf.tests):
+@pytest.mark.parametrize(
+    "i,test,udf",
+    [
+        (i, test, udf)
+        for udf in parse_udf.parse_udf_dir('udf')
+        for i, test in enumerate(udf.tests)
+    ]
+)
+def test_udfs(bq, i, test, udf):
```

**Comment by jklukas** [comment]
Out of draft, so ready for another look.

I've added test-specific udfs for making asserts, which will hopefully make tests easier to read and write.

File: udf/udf_country_code_to_flag.sql [hunk 3]
```diff
@@ -22,10 +11,19 @@ See:
...
+      SELECT
+        -- This constant offset transforms from ASCII characters A-Z to
+        -- the equivalent Unicode regional indicator symbols A-Z.
+        -- See https://en.wikipedia.org/wiki/Regional_Indicator_Symbol
+        c + 127397
+      FROM
+        UNNEST(TO_CODE_POINTS(country_code)) c)));
 
-*/
+SELECT
```
Comment (relud) [inline udf/udf_country_code_to_flag.sql]:
```suggestion
-- Tests

SELECT
```

File: udf/udf_bitmask_range.sql [hunk 4]
```diff
@@ -16,16 +8,18 @@ The length argument is the number of bits to include in the mask.
...
+CREATE TEMP FUNCTION
+  udf_bitmask_range( start_ordinal INT64,
+    _length INT64) AS ((
+    SELECT
+      SUM(1 << (_n - 1))
+    FROM
+      UNNEST(GENERATE_ARRAY(start_ordinal, start_ordinal + _length - 1)) AS _n ));
 
-*/
+SELECT
```
Comment (relud) [inline udf/udf_bitmask_range.sql]:
```suggestion
-- Tests

SELECT
```

**Pull request merged**